File: constructs/pulumi/go/network.go
// Package infrapulumi provides Pulumi component wrappers for the Nautilus
// platform Terraform modules. Resource creation is delegated entirely to
// Terraform via Pulumi-Terraform module interop; no Azure resources are
// created directly by this package.
package infrapulumi

import (
	"fmt"

	"github.com/nautilus/infra-pulumi-go/policy"
	tfm "github.com/pulumi/pulumi-terraform-module-sdk/go/terraformmodule"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

const (
	moduleRepo    = "git::ssh://git@github.com/nautilus/terraform-modules.git"
	moduleVersion = "v1.0.0"
	networkSource = moduleRepo + "//modules/networking?ref=" + moduleVersion
)

var validEnvironments = map[string]bool{"dev": true, "staging": true, "prod": true}

// SubnetDelegation holds service delegation configuration for a subnet.
type SubnetDelegation struct {
	Name    *string
	Service *string
	Actions *[]*string
}

// SubnetConfig holds configuration for a single subnet within the VNet.
type SubnetConfig struct {
	AddressPrefix    *string
	ServiceEndpoints *[]*string
	Delegation       *SubnetDelegation
	NsgRules         *[]map[string]interface{}
}

// NetworkComponentArgs holds all inputs for NetworkComponent.
type NetworkComponentArgs struct {
	Project         *string
	Environment     *string
	ResourceGroup   *string
	Location        *string
	AddressSpace    *[]*string
	Subnets         *map[string]*SubnetConfig
	PrivateDnsZones *[]*string
	ExtraTags       *map[string]*string
}

// NetworkComponent provisions a VNet with subnets, NSGs, and private DNS zones.
//
// Delegates to the platform modules/networking Terraform module via
// Pulumi-Terraform interop. Terraform creates every Azure resource;
// Pulumi reads the outputs.
type NetworkComponent struct {
	pulumi.ResourceState

	// VnetId is the resource ID of the VNet.
	VnetId pulumi.StringOutput
	// VnetName is the name of the VNet.
	VnetName pulumi.StringOutput
	// SubnetIds is a map of subnet name → resource ID.
	SubnetIds pulumi.Output
	// DnsZoneIds is a map of DNS zone name → resource ID.
	DnsZoneIds pulumi.Output
}

// NewNetworkComponent creates a new NetworkComponent.
func NewNetworkComponent(
	ctx *pulumi.Context,
	name string,
	args *NetworkComponentArgs,
	opts ...pulumi.ResourceOption,
) (*NetworkComponent, error) {
	env := *args.Environment
	if !validEnvironments[env] {
		return nil, fmt.Errorf("environment must be one of [dev, prod, staging], got %q", env)
	}

	comp := &NetworkComponent{}
	if err := ctx.RegisterComponentResource("nautilus:network:NetworkComponent", name, comp, opts...); err != nil {
		return nil, err
	}

	tags := policy.RequiredTags(args.Project, args.Environment, args.ExtraTags)

	// Serialise subnets to map[string]interface{} for the TF module
	subnetsVar := map[string]interface{}{}
	if args.Subnets != nil {
		for subnetName, cfg := range *args.Subnets {
			entry := map[string]interface{}{"address_prefix": *cfg.AddressPrefix}
			if cfg.ServiceEndpoints != nil {
				entry["service_endpoints"] = *cfg.ServiceEndpoints
			}
			if cfg.Delegation != nil {
				entry["delegation"] = map[string]interface{}{
					"name":    *cfg.Delegation.Name,
					"service": *cfg.Delegation.Service,
					"actions": *cfg.Delegation.Actions,
				}
			}
			if cfg.NsgRules != nil {
				entry["nsg_rules"] = *cfg.NsgRules
			}
			subnetsVar[subnetName] = entry
		}
	}

	privateDnsZones := []string{}
	if args.PrivateDnsZones != nil {
		for _, z := range *args.PrivateDnsZones {
			privateDnsZones = append(privateDnsZones, *z)
		}
	}

	mod, err := tfm.NewModule(ctx, name+"-networking", &tfm.ModuleArgs{
		Source: pulumi.String(networkSource),
		Variables: pulumi.Map{
			"project":             pulumi.String(*args.Project),
			"environment":         pulumi.String(*args.Environment),
			"resource_group_name": pulumi.String(*args.ResourceGroup),
			"location":            pulumi.String(*args.Location),
			"address_space":       pulumi.ToStringArray(*args.AddressSpace),
			"subnets":             pulumi.ToMap(subnetsVar),
			"private_dns_zones":   pulumi.ToStringArray(privateDnsZones),
			"tags":                pulumi.ToStringMap(*tags),
		},
	}, pulumi.Parent(comp))
	if err != nil {
		return nil, err
	}

	comp.VnetId     = mod.GetOutput("vnet_id").(pulumi.StringOutput)
	comp.VnetName   = mod.GetOutput("vnet_name").(pulumi.StringOutput)
	comp.SubnetIds  = mod.GetOutput("subnet_ids")
	comp.DnsZoneIds = mod.GetOutput("dns_zone_ids")

	if err := ctx.RegisterResourceOutputs(comp, pulumi.Map{
		"vnetId":     comp.VnetId,
		"vnetName":   comp.VnetName,
		"subnetIds":  comp.SubnetIds,
		"dnsZoneIds": comp.DnsZoneIds,
	}); err != nil {
		return nil, err
	}

	return comp, nil
}

File: constructs/pulumi/typescript/src/network.ts
import * as pulumi from "@pulumi/pulumi";
import * as tfm from "@pulumi/terraform-module";
import { requiredTags } from "./policy/tagging";

const VALID_ENVIRONMENTS = new Set(["dev", "staging", "prod"]);

const MODULE_REPO    = "git::ssh://git@github.com/nautilus/terraform-modules.git";
const MODULE_VERSION = "v1.0.0";
const NETWORK_SOURCE = `${MODULE_REPO}//modules/networking?ref=${MODULE_VERSION}`;

export interface SubnetDelegation {
  readonly name:    string;
  readonly service: string;
  readonly actions: string[];
}

export interface SubnetConfig {
  readonly addressPrefix:     string;
  readonly serviceEndpoints?: string[];
  readonly delegation?:       SubnetDelegation;
  readonly nsgRules?:         Record<string, unknown>[];
}

export interface NetworkComponentArgs {
  readonly project:          string;
  readonly environment:      string;
  readonly resourceGroup:    string;
  readonly location:         string;
  readonly addressSpace:     string[];
  readonly subnets?:         Record<string, SubnetConfig>;
  readonly privateDnsZones?: string[];
  readonly extraTags?:       Record<string, string>;
}

function serialiseSubnets(subnets: Record<string, SubnetConfig>): Record<string, unknown> {
  const result: Record<string, unknown> = {};
  for (const [name, cfg] of Object.entries(subnets)) {
    const entry: Record<string, unknown> = { address_prefix: cfg.addressPrefix };
    if (cfg.serviceEndpoints?.length) entry["service_endpoints"] = cfg.serviceEndpoints;
    if (cfg.delegation) {
      entry["delegation"] = {
        name:    cfg.delegation.name,
        service: cfg.delegation.service,
        actions: cfg.delegation.actions,
      };
    }
    if (cfg.nsgRules?.length) entry["nsg_rules"] = cfg.nsgRules;
    result[name] = entry;
  }
  return result;
}

/**
 * NetworkComponent provisions a VNet with subnets, NSGs, and private DNS zones.
 *
 * Delegates to the platform `modules/networking` Terraform module via
 * Pulumi-Terraform interop. Terraform creates every Azure resource;
 * Pulumi reads the outputs.
 */
export class NetworkComponent extends pulumi.ComponentResource {
  /** Resource ID of the VNet. */
  public readonly vnetId: pulumi.Output<string>;
  /** Name of the VNet. */
  public readonly vnetName: pulumi.Output<string>;
  /** Map of subnet name → resource ID. */
  public readonly subnetIds: pulumi.Output<Record<string, string>>;
  /** Map of DNS zone name → resource ID. */
  public readonly dnsZoneIds: pulumi.Output<Record<string, string>>;

  constructor(name: string, args: NetworkComponentArgs, opts?: pulumi.ComponentResourceOptions) {
    if (!VALID_ENVIRONMENTS.has(args.environment)) {
      throw new Error(
        `environment must be one of [${[...VALID_ENVIRONMENTS].sort().join(", ")}], got "${args.environment}"`
      );
    }

    super("nautilus:network:NetworkComponent", name, {}, opts);

    const tags = requiredTags(args.project, args.environment, args.extraTags);

    const mod = new tfm.Module(`${name}-networking`, {
      source: NETWORK_SOURCE,
      variables: {
        project:             args.project,
        environment:         args.environment,
        resource_group_name: args.resourceGroup,
        location:            args.location,
        address_space:       args.addressSpace,
        subnets:             serialiseSubnets(args.subnets ?? {}),
        private_dns_zones:   args.privateDnsZones ?? [],
        tags,
      },
    }, { parent: this });

    this.vnetId     = mod.getOutput("vnet_id");
    this.vnetName   = mod.getOutput("vnet_name");
    this.subnetIds  = mod.getOutput("subnet_ids");
    this.dnsZoneIds = mod.getOutput("dns_zone_ids");

    this.registerOutputs({
      vnetId:     this.vnetId,
      vnetName:   this.vnetName,
      subnetIds:  this.subnetIds,
      dnsZoneIds: this.dnsZoneIds,
    });
  }
}
